File: solved/377.CombinationSumIV.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

map<int, int> memoization;

int recursion(vector<int>& nums, int target) {
  if (memoization.find(target) != memoization.end()) {
    return memoization[target];
  }

  if (target == 0) {
    memoization[target] = 1;
    return 1;
  }

  int res = 0;
  for (int i = 0; i < nums.size(); i++) {
    if (target >= nums[i]) {
      res += recursion(nums, target - nums[i]);
    }
  }

  memoization[target] = res;
  return res;
}

class Solution {
public:
  int combinationSum4(vector<int>& nums, int target) {
    memoization.clear();
    int res = recursion(nums, target);
    cout << res << endl;
    return res;
  }
};

int main() {
  Solution s;
  vector<int> nums = {1, 2, 3};
  s.combinationSum4(nums, 32);
}

File: solved/377.CombinationSumIV.py
memoization = {}

def recursion(nums, target):
  if target in memoization:
    return memoization[target]

  if target == 0:
    memoization[target] = 1
    return 1
  
  count = 0
  for n in nums:
    if n <= target: 
      count += recursion(nums, target - n)
  
  memoization[target] = count
  return count

class Solution(object):
  def combinationSum4(self, nums, target):
    """
    :type nums: List[int]
    :type target: int
    :rtype: int
    """
    memoization.clear()
    res = recursion(nums, target)
    print(res)
    return res

Solution.combinationSum4(0, [1,2,3], 4)
Solution.combinationSum4(0, [9], 3)
Solution.combinationSum4(0, [1,2,3], 32)
